# Fundamentos de Redes Neurais

## Objetivo da Aula
Nesta aula, vamos explorar os **fundamentos de redes neurais**, com foco no funcionamento de um **único neurônio artificial**. Vamos explicar passo a passo o código fornecido, entendendo cada linha e conceito envolvido.

## Introdução às Redes Neurais

As **redes neurais artificiais** são modelos computacionais inspirados no cérebro humano. Elas são compostas por unidades chamadas **neurônios artificiais**, que recebem entradas, processam informações e produzem uma saída. Um neurônio pode ser descrito como:

1. Receber um conjunto de valores de entrada.
2. Multiplicar cada entrada pelo seu respectivo peso (importância).
3. Somar esses produtos e adicionar um valor constante chamado *bias* (ou viés).
4. Aplicar uma **função de ativação** para determinar a saída final do neurônio.

Vamos analisar isso detalhadamente com o código abaixo.

## Código Explicado Passo a Passo

### 1. Importando Bibliotecas

In [1]:
import numpy as np

- **`numpy`** é uma biblioteca usada para operações matemáticas em arrays multidimensionais.
- Aqui, ela será usada para calcular o produto escalar entre as entradas e os pesos.

### 2. Definindo a Função do Neurônio

In [2]:
def neuron_output(inputs, weights, bias, activation='relu'):
    """
    Simula um único neurônio:
      - inputs: array np (features de entrada)
      - weights: array np (pesos)
      - bias: valor escalar
      - activation: 'relu' ou 'sigmoid' (demonstrativo)
    """
    # Soma ponderada
    z = np.dot(inputs, weights) + bias
    
    # Aplica função de ativação
    if activation == 'relu':
        return max(0, z)
    elif activation == 'sigmoid':
        return 1.0 / (1.0 + np.exp(-z))
    else:
        # Se não reconhece a ativação, retorna z sem ativação
        return z

```python
def neuron_output(inputs, weights, bias, activation='relu'):
```

- Esta função simula o comportamento de **um único neurônio artificial**.
- Parâmetros:
  - `inputs`: vetor de números reais representando os dados de entrada.
  - `weights`: vetor com os pesos associados às entradas.
  - `bias`: valor escalar que ajusta a saída do neurônio.
  - `activation`: tipo de função de ativação utilizada (`'relu'` ou `'sigmoid'`).


**Cálculo da Soma Ponderada**

```python
z = np.dot(inputs, weights) + bias
```

- **`np.dot(inputs, weights)`**: calcula o **produto escalar** entre os vetores `inputs` e `weights`.
    - Isso equivale a somar o produto de cada entrada com seu peso correspondente:  
      $ z = x_1w_1 + x_2w_2 + \cdots + x_nw_n $
- Depois, adicionamos o **viés (`bias`)** à soma.
- O resultado é uma única saída numérica chamada `z`.

**ReLU (Rectified Linear Unit)**

```python
if activation == 'relu':
    return max(0, z)
```

- A função **ReLU** retorna zero se a entrada for negativa, e a própria entrada se for positiva.
- Matematicamente:  
  $$
  f(z) = \max(0, z)
  $$
- É muito usada em redes neurais modernas por sua simplicidade e eficiência na propagação de gradientes.

**Sigmoid**

```python
elif activation == 'sigmoid':
    return 1.0 / (1.0 + np.exp(-z))
```

- A função **Sigmoid** transforma qualquer número real em um valor entre 0 e 1.
- Útil para problemas de classificação binária ou para representar probabilidades.
- Fórmula:  
  $$
  f(z) = \frac{1}{1 + e^{-z}}
  $$


**Sem Ativação**

```python
else:
    return z
```

- Se nenhuma das funções acima for especificada, a função retorna apenas o valor de `z`, ou seja, sem aplicar função de ativação.

## Exemplo de Uso do Código

In [3]:
inputs = np.array([0.5, -0.2, 1.0])   # 3 entradas
weights = np.array([0.8, -0.5, 0.1]) # 3 pesos correspondentes
bias = 0.2

- Temos 3 entradas: `[0.5, -0.2, 1.0]`
- Três pesos: `[0.8, -0.5, 0.1]`
- Um valor de *bias*: `0.2`

### Chamando a Função

In [4]:
out_relu = neuron_output(inputs, weights, bias, 'relu')
out_sigmoid = neuron_output(inputs, weights, bias, 'sigmoid')

- Calculamos a saída do neurônio com ambas as funções de ativação.

### Saída do Programa

In [5]:
print("=== Momento 2: Fundamentos de Redes Neurais ===")
print("Inputs :", inputs)
print("Weights:", weights)
print("Bias   :", bias)
print("Saída (ReLU)   :", out_relu)
print("Saída (Sigmoid):", out_sigmoid)

=== Momento 2: Fundamentos de Redes Neurais ===
Inputs : [ 0.5 -0.2  1. ]
Weights: [ 0.8 -0.5  0.1]
Bias   : 0.2
Saída (ReLU)   : 0.8
Saída (Sigmoid): 0.6899744811276125


## Entendendo os Resultados

### Cálculo da Soma Ponderada:

$$
z = (0.5 \cdot 0.8) + (-0.2 \cdot -0.5) + (1.0 \cdot 0.1) + 0.2 = \\
= 0.4 + 0.1 + 0.1 + 0.2 = 0.8
$$

### Aplicando as Funções:

- **ReLU**: `max(0, 0.8) = 0.8`
- **Sigmoid**: $ \frac{1}{1 + e^{-0.8}} \approx 0.69 $

(Os valores podem variar levemente dependendo do arredondamento.)

## Resumo dos Conceitos Abordados

| Conceito | Descrição |
|---------|-----------|
| **Neurônio Artificial** | Unidade básica de uma rede neural, que recebe entradas, multiplica por pesos e gera uma saída. |
| **Pesos** | Valores ajustáveis que definem a importância de cada entrada. |
| **Bias** | Termo constante que desloca a saída do neurônio. |
| **Produto Escalar** | Operação que combina entradas e pesos. |
| **Função de Ativação** | Determina a forma como o neurônio "responde" à entrada. |
| **ReLU** | Função simples e popular: retorna zero para valores negativos. |
| **Sigmoid** | Mapeia qualquer valor para um intervalo entre 0 e 1. Útil em classificações. |

## Conclusão

- Com este exemplo prático, vimos como funciona um **único neurônio artificial**. 
- Este é o bloco fundamental de qualquer rede neural profunda. 
- Em aulas futuras, vamos conectar múltiplos neurônios em camadas e construir redes neurais completas capazes de resolver problemas complexos.

## Atividades Propostas


1. Modifique os valores de `inputs`, `weights` e `bias` e veja como as saídas mudam.
1. Calcule manualmente o valor de `z` e compare com o valor retornado pela função.

